### ISSUE TITLE
Server Islands script execution only once

### ISSUE THREAD

**Issue by JeroenReumkens** [withastro/astro#12294]
### Astro Info

```block
v4.16.7
Node                     v18.17.1
System                   macOS (arm64)
Package Manager          unknown
Output                   server
Adapter                  @astrojs/vercel/serverless
Integrations             @astrojs/tailwind
                         astro-expressive-code
                         @astrojs/mdx
                         @astro-clerk-auth/integration
                         @astrojs/react
                         @sentry/astro
```

### Describe the Bug

Using `server islands` and `directRenderScript`, I noticed that scripts are injected multiple times but only executed once. I expected `directRenderScript` to ensure the script runs every time it is injected, but since the same file is reused, it doesn’t.

Is this a bug, or are there other lifecycle events in JavaScript that I can use to rerun scripts when new deferred elements are added? This is especially important for addEventListener scripts.

In the Stackblitz example below, a querySelector changes a div’s background color, but it only works for the first loaded element, not the second. I could use MutationObserver to manually rerun the script, but that feels too heavy for simple event listeners.

### Real-world use case

I’m making my Mux video player a server:defer component. When it loads, I bind an event to set the playback speed to a previously stored value from localStorage.

### What's the expected result?

I expect that either the script runs for every component that gets injected, or there is some sort of event I can tap into to rerun the script myself. 

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-lgq5uz?file=src%2Fpages%2Findex.astro

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by JeroenReumkens** [comment]
To anyone running into this issue to: 

I just got a brain fart when pouring my morning espresso. I added `define:vars` to the script tag passing in a unique ID. That way I can query for this specific element only, and only bind events that haven't yet. `define:vars` also implicitly sets `is:inline`, meaning that the script is unique per component, and inline added for every component being server deferred. 

Downside is that this code now is duplicated, but for the time being that is a good enough solution for me. 
Looking forward to hearing if there is a different approach to solve this.

**Comment by alexgipi** [comment]
This also happens with ViewTransitions. When navigating between pages, the server island script doesn’t work—it only functions on the initial page load. Adding an event listener for the astro:after-swap event doesn’t seem to have any effect. I’m not sure if this is the expected behavior or if I might be missing something!

**Comment by MaciejGarncarski** [comment]
It's still hapening with View Transitions. If I remove `<ClientRouter/>` it works fine.

**Comment by MaciejGarncarski** [comment]
I found a workaround. In my example, I have a blog post with views, which are server island. The disadvantage of this workaround is fetching sever island 2 times, first time at render and second time when this script runs. I put this script on layout page, may vary based on your needs.

```ts
<script>
  const remakeScript = async () => {
    await new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve(null);
      }, 300);
    }); // can be deleted

    const viewsElement = document.querySelector("[data-views]");

    if (viewsElement) {
      return;
    }

    const islandScript = document.querySelector("[data-island-id]");
    const scriptAtributes = islandScript?.attributes;

    const copiedIslandScript = islandScript?.cloneNode(true);

    const newScript = document.createElement("script");

    [...Array.from(scriptAtributes || [])].forEach(({ name, value }) =>
      newScript.setAttribute(name, value)
    );

    newScript.textContent = `
async function workaround(){
  ${copiedIslandScript?.textContent || ""}
  return "workaround completed"
}

workaround()
    `;

    islandScript?.remove();
    islandScript?.parentElement?.appendChild(newScript);
  };

  document.addEventListener("astro:after-swap", remakeScript);
</script>
```

**Comment by matthewp** [comment]
We're closing stale issues to keep the backlog manageable as the project has grown. Since this issue hasn't seen activity in a while, we're closing it out. If you're still experiencing this in Astro 6, please open a new issue and we'll take a fresh look. Thank you for reporting it!
